docs/docs/04_docker_intro/04_in_depth_docker_tutorial/README.md

```diff
@@ -0,0 +1,236 @@
+  -w /app -v "$(pwd):/app" \
+  --network network-name \
+  -e MYSQL_HOST=mysql \
+  -e MYSQL_USER=root \
+  -e MYSQL_PASSWORD=secret \
+  -e MYSQL_DB=todos \
+  node:12-alpine \
+  sh -c "npm install && npm run dev"
+```
+
+## How to run multiple containers using Docker Compose
+
+1. Create a `docker-compose.yml` file in the root of your project.
+2. Turn each of the `docker run` commands into a `service` in the `docker-compose.yml` file.
+3. This is basically re-creating the flags passed to the `docker run` command, but in YAML format.
+
+Example of the two `docker run`s above:
+
+```yml
+services:
+  app:
+    image: node:12-alpine
+    command: sh -c "npm install && npm run dev"
+    ports:
+      - 3000:3000
+    working_dir: /app
+    volumes:
+      - ./:/app
+    environment:
+      MYSQL_HOST: mysql
+      MYSQL_USER: root
+      MYSQL_PASSWORD: secret
+      MYSQL_DB: todos
+  mysql:
+    image: mysql:5.7
+    platform: linux/amd64
+    volumes:
+      - todo-mysql-data:/var/lib/mysql
+    environment:
+      MYSQL_ROOT_PASSWORD: secret
+      MYSQL_DATABASE: todos
+
+volumes:
+  todo-mysql-data:
+```
+
+Then, just run `docker compose up -d` and it will start in the background!
+
+You can see it in Docker desktop.
+
+Tear it down and remove the containers (but not the volumes) with `docker compose down`.
+
+## Caching in Dockerfile layers
+
+Each layer (basically line of text) in a Dockerfile uses caching.
+
+That means that if Docker doesn't detect that a layer has changed, it won't re-run it. It'll use the last value / files that were generated for the last build.
+
+However, it also means that if one layer changes and has to be re-built, Docker will re-build all subsequent layers.
+
+Therefore it's best to set up your Dockerfile so that you can maximise the amount of caching and reduce the chances of a cache bust.
+
+For example, instead of this:
+
+```
+FROM node:12-alpine
+WORKDIR /app
+COPY . .
+RUN npm install --production
+CMD ["node", "src/index.js"]
+```
+
+You might do this:
+
+```
+FROM node:12-alpine
+WORKDIR /app
+COPY package.json package.lock ./
+RUN npm install --production
+COPY . .
+CMD ["node", "src/index.js"]
+```
+
+That way if the `package.json` and `package.lock` files don't change, you won't re-run `npm install`.
+
+In the first example, if _any_ code files changed, `npm install` would run. Even if it was not needed because the requirements file didn't change.
+
+### Ignore certain files and folders with `.dockerignore`
+
+Some files and folders can be safely ignored when copying over to the Docker container. For example, `node_modules` or the Python virtual environment.
+
+Create a `.dockerignore` file in the root directory of your project (where `docker-compose.yml` lives), and add this (more examples of what to add for a Python project [here](https://github.com/GoogleCloudPlatform/getting-started-python/blob/main/optional-kubernetes-engine/.dockerignore)):
+
+```
+node_modules
+.venv
+*.pyc
+__pycache__
+```
```

Comment:
I think Docker secrets should really be named since the variables passed in the Dockerfile will be saved in the image when built.
[Click for more info](https://docs.docker.com/engine/swarm/secrets/)

Reply:
I would mention that when used in production It is a good practice to manage sensitive data using secrets. 

Additionally, I would warn the users that if they plan to share the created images, they should avoid passing sensitive data in the environment variables and consider using docker secrets instead. Even though, this approach requires using docker swarm.

Reply:
Makes sense, those are good points! I will include them. Thanks 🙏

Reply:
I've added the two warnings:

- https://github.com/tecladocode/rest-apis-flask-python/blob/jose/cou-17-write-docker-section/docs/docs/04_docker_intro/04_in_depth_docker_tutorial/README.md#how-to-pass-environment-variables-to-a-container
- https://github.com/tecladocode/rest-apis-flask-python/blob/jose/cou-17-write-docker-section/docs/docs/04_docker_intro/04_in_depth_docker_tutorial/README.md#how-to-pass-environment-variables-to-a-container